integration/querydsl/blaze-persistence-querydsl-expressions/src/main/java/com/blazebit/persistence/querydsl/AbstractBlazeJPAQuery.java

```diff
@@ -0,0 +1,527 @@
+
+        for (Map.Entry<String, Object> entry : hints.entries()) {
+            try {
+                query.setHint(entry.getKey(), entry.getValue());
+            } catch (UnsupportedOperationException e) {
+                LOG.warning("Failed to set query hint");
+            }
+        }
+
+        logQuery(criteriaBuilder.getQueryString(), null);
+        return query;
+    }
+
+    @Override
+    public PagedList<T> fetchPage(int firstResult, int maxResults) {
+        return  getFullQueryBuilder(null)
+                .page(firstResult, maxResults)
+                .getResultList();
+    }
+
+    @Override
+    public PagedList<T> fetchPage(KeysetPage keysetPage, int firstResult, int maxResults) {
+        return getFullQueryBuilder(null)
+                .page(keysetPage, firstResult, maxResults)
+                .getResultList();
+    }
+
+    public String getQueryString() {
+        return getQueryable(null).getQueryString();
+    }
+
+    protected FullQueryBuilder<T, ?> getFullQueryBuilder(@Nullable QueryModifiers modifiers) {
+        Queryable<T, ?> queryable = getQueryable(modifiers);
+        if (queryable instanceof FullQueryBuilder) {
+            return (FullQueryBuilder<T, ?>) queryable;
+        }
+        throw new UnsupportedOperationException("This feature is not yet supported on " + queryable.getClass().getSimpleName());
+    }
+
+    protected Queryable<T, ?> getQueryable(@Nullable QueryModifiers modifiers) {
+        BlazeCriteriaBuilderRenderer<T> blazeCriteriaBuilderRenderer = new BlazeCriteriaBuilderRenderer<T>(criteriaBuilderFactory, entityManager, getTemplates());
+        Queryable<T, ?> queryable = blazeCriteriaBuilderRenderer.render(this);
+        CriteriaBuilder<T> criteriaBuilder = blazeCriteriaBuilderRenderer.getCriteriaBuilder();
+
+        if (modifiers != null) {
+            if (modifiers.getLimitAsInteger() != null) {
+                criteriaBuilder.setMaxResults(modifiers.getLimitAsInteger());
+            }
+            if (modifiers.getOffsetAsInteger() != null) {
+                criteriaBuilder.setFirstResult(modifiers.getOffsetAsInteger());
+            }
+        }
+
+        for (Map.Entry<String, Object> entry : hints.entries()) {
+            criteriaBuilder.setProperty(entry.getKey(), entry.getValue().toString());
+        }
+
+        if (cachable) {
+            criteriaBuilder.setCacheable(true);
+        }
+
+        return queryable;
+    }
+
+    @Override
+    protected void clone(Q query) {
+        super.clone(query);
+        this.cachable = query.cachable;
+    }
+
+    // Work around private access to query(...)
+
+    @Override
+    public long fetchCount() {
+        try {
+            Query query = createQuery(null, true);
+            return (Long) query.getSingleResult();
+        } finally {
+            reset();
+        }
+    }
+
+    @Override
+    public QueryResults<T> fetchResults() {
+        try {
+            Query countQuery = createQuery(null, true);
```

Comment:
`createQuery`(..., true : forCount)`  returns the `TypedQuery` from `getFullQueryBuilder(...).getCountQuery()` under the hood.

Reply:
`PaginatedCriteriaBuilder.getTypedQuery()` returns a `PaginatedTypedQuery` which then returns the `PagedList` from `getResultList()` which is much more powerful than invoking a separate count query like you do here.

Reply:
Ah I only now see what this method is doing..

Usually one uses `fetch()` wich is equivalent to our `getResultList()`. There is also `fetchOne()` (`getSingleResult`) and `fetchFirst()`.

Because implementing `Queryable` would only throw more confusing methods in the game, I did not do that. But I did port some of the pagination API's over to `BlazeJPAQuery`.

The `fetchResults` method seems to be some sort of mechanism to perform the count query first (without any limit or offset), and only perform the query with projection if there is any result (and use limit and offset in that case).

I could add something like:

```
QueryModifiers modifiers = getMetadata().getModifiers();
Long offset = modifiers.getOffset();
Long limit = modifiers.getLimit();

if (offset != null && limit != null) {
    Queryable<T, ?> queryable = getQueryable(null);
    if (queryable instanceof FullQueryBuilder) {
        FullQueryBuilder<T, ?> fullQueryBuilder = (FullQueryBuilder<T, ?>) queryable;
        PaginatedCriteriaBuilder<T> page = fullQueryBuilder.page(offset.intValue(), limit.intValue());
        PagedList<T> resultList = page.getResultList();
        return new QueryResults<>(resultList, modifiers, resultList.getTotalSize());
    }
} else {
    // Perform the old code.
}
```

But this wouldn't prevent users from using this in cases where pagination is not allowed, for example when no viable sorting is used).

In reality, I think `fetchResults`  should be discouraged in favour for the actual pagination API methods.

I however had to override it, because it calls `createQuery` internally, which is a private method. So rather than overriding `createQuery`, I had to "copy" all methods that use it over so that they call my extended version of `createQuery` that is actually able to render JPQL.Next queries... My PR to open `createQuery` up for extension actually got merged ( https://github.com/querydsl/querydsl/pull/2517 ) but we can't rely on this change being available or we would have to force our users to be on the absolute newest version of Querydsl.

Perhaps I should just leave the code as is, but throw a warning on the logger indicating that the pagination methods are way more performant for this type of use case?

Reply:
I'm not sure I understand what you mean here. The `PaginatedCriteriaBuilder` API does the same i.e. first execute a count query and then the main query with limit/offset if the count is > 0.
The `PaginatedCriteriaBuilder` API additionally allows to inline the count query into the id or main query to avoid multiple queries, which is the default, but can be turned off.